pkg/vere/mdns.c

```diff
@@ -0,0 +1,220 @@
+    c3_free(payload);
+    c3_free(req);
+    DNSServiceRefDeallocate(sref);
+  }
+}
+
+static void poll_cb(uv_poll_t *handle, int status, int events) {
+  DNSServiceRef sref = (DNSServiceRef) handle->data;
+  int err = DNSServiceProcessResult(sref);
+}
+
+static void dns_sd_cb(DNSServiceRef sref, mdns_payload* payload) {
+  int fd = DNSServiceRefSockFD(sref);
+  uv_loop_t* loop = uv_default_loop();
+  uv_poll_t* poll = (uv_poll_t *)c3_malloc(sizeof(uv_poll_t));
+  poll->data = (void *)sref;
+  payload->poll = poll;
+  uv_poll_init(loop, poll, fd);
+  uv_poll_start(poll, UV_READABLE, poll_cb);
+}
+
+static void browse_cb(DNSServiceRef s,
+                      DNSServiceFlags f,
+                      uint32_t interface,
+                      DNSServiceErrorType err,
+                      const char* name,
+                      const char* type,
+                      const char* domain,
+                      void* context)
+{
+  if (err != kDNSServiceErr_NoError) {
+    u3l_log("mdns: service browse error %i", err);
+    return;
+  }
+
+  if (f & kDNSServiceFlagsAdd) {	// Add
+    DNSServiceRef sr;
+
+    mdns_payload* payload = (mdns_payload*)context;
+    mdns_payload* payload_copy = c3_malloc(sizeof *payload_copy);
+
+    // copy to prevent asynchronous thrashing of payload
+    memcpy(payload_copy, payload, sizeof(mdns_payload));
+
+    DNSServiceErrorType err =
+      DNSServiceResolve(&sr, 0, interface,
+                        name, type, domain, resolve_cb,
+                        (void*)payload_copy);
+
+    if (err != kDNSServiceErr_NoError) {
+      u3l_log("mdns: dns service resolve error %i", err);
+      c3_free(payload_copy);
+      DNSServiceRefDeallocate(sr);
+      return;
+    }
+    dns_sd_cb(sr, payload_copy);
+  }
+}
+
+static void register_cb(DNSServiceRef sref,
+                        DNSServiceFlags f,
+                        DNSServiceErrorType err,
+                        const char* name,
+                        const char* type,
+                        const char* domain,
+                        void* context)
+{
+  mdns_payload* payload = (mdns_payload*)context;
+
+  if (err != kDNSServiceErr_NoError) {
+    u3l_log("mdns: service register error %i", err);
+  } else {
+    u3l_log("mdns: %s registered on all interfaces", name);
+  }
+  uv_poll_stop(payload->poll);
+  c3_free(payload->poll);
+  c3_free(payload->our);
+  c3_free(payload);
+}
+
+void mdns_init(uint16_t port, char* our, mdns_cb* cb, void* context)
+{
+  #if defined(U3_OS_linux)
+  setenv("AVAHI_COMPAT_NOWARN", "1", 0);
+  setenv("DBUS_SYSTEM_BUS_ADDRESS", "unix:path=/var/run/dbus/system_bus_socket", 0);
+  #   endif
+
+  memmove(our, our+1, strlen(our)); // certain url parsers don't like the sig
+
+  mdns_payload* register_payload = (mdns_payload*)c3_malloc(sizeof(mdns_payload));
+
+  register_payload->cb = cb;
+  register_payload->context = context;
+  register_payload->our = our;
+
```

Comment:
It looks like you don't need to save `our` in `payload`. I would try removing that, and also just passing `our + 1` to `DNSServiceRegister()` (maybe after checking that `'~' = *our`).

Reply:
Done